# Ex1 - Getting and Knowing your Data



### Step 1. Import the necessary libraries

In [1]:
import numpy as np
import pandas as pd

### Step 2. Import the dataset

In [2]:
pd.read_csv("D:/Ad_py_lib.csv/pandas_and_numpy/pandas_data.csv")

order_id  quantity                              item_name  \
0            1         1           Chips and Fresh Tomato Salsa   
1            1         1                                   Izze   
2            1         1                       Nantucket Nectar   
3            1         1  Chips and Tomatillo-Green Chili Salsa   
4            2         2                           Chicken Bowl   
...        ...       ...                                    ...   
4617      1833         1                          Steak Burrito   
4618      1833         1                          Steak Burrito   
4619      1834         1                     Chicken Salad Bowl   
4620      1834         1                     Chicken Salad Bowl   
4621      1834         1                     Chicken Salad Bowl   

                                     choice_description item_price  
0                                                   NaN     $2.39   
1                                          [Clementine]     $3.39   
2                                               [Apple]     $3.39   
3                                                   NaN     $2.39   
4     [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98   
...                                                 ...        ...  
4617  [Fresh Tomato Salsa, [Rice, Black Beans, Sour ...    $11.75   
4618  [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese...    $11.75   
4619  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...    $11.25   
4620  [Fresh Tomato Salsa, [Fajita Vegetables, Lettu...     $8.75   
4621  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...     $8.75   

[4622 rows x 5 columns]

### Step 3. Assign it to a variable called a.

In [3]:
a = pd.read_csv("D:/Ad_py_lib.csv/pandas_and_numpy/pandas_data.csv")
a

order_id  quantity                              item_name  \
0            1         1           Chips and Fresh Tomato Salsa   
1            1         1                                   Izze   
2            1         1                       Nantucket Nectar   
3            1         1  Chips and Tomatillo-Green Chili Salsa   
4            2         2                           Chicken Bowl   
...        ...       ...                                    ...   
4617      1833         1                          Steak Burrito   
4618      1833         1                          Steak Burrito   
4619      1834         1                     Chicken Salad Bowl   
4620      1834         1                     Chicken Salad Bowl   
4621      1834         1                     Chicken Salad Bowl   

                                     choice_description item_price  
0                                                   NaN     $2.39   
1                                          [Clementine]     $3.39   
2                                               [Apple]     $3.39   
3                                                   NaN     $2.39   
4     [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98   
...                                                 ...        ...  
4617  [Fresh Tomato Salsa, [Rice, Black Beans, Sour ...    $11.75   
4618  [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese...    $11.75   
4619  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...    $11.25   
4620  [Fresh Tomato Salsa, [Fajita Vegetables, Lettu...     $8.75   
4621  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...     $8.75   

[4622 rows x 5 columns]

In [31]:
# a = pd.read_csv("D:/Ad_py_lib.csv/pandas_and_numpy/pandas_data.csv",index_col = "order_id")
a

order_id  quantity                              item_name  \
0            1         1           Chips and Fresh Tomato Salsa   
1            1         1                                   Izze   
2            1         1                       Nantucket Nectar   
3            1         1  Chips and Tomatillo-Green Chili Salsa   
4            2         2                           Chicken Bowl   
...        ...       ...                                    ...   
4617      1833         1                          Steak Burrito   
4618      1833         1                          Steak Burrito   
4619      1834         1                     Chicken Salad Bowl   
4620      1834         1                     Chicken Salad Bowl   
4621      1834         1                     Chicken Salad Bowl   

                                     choice_description item_price  
0                                                   NaN     $2.39   
1                                          [Clementine]     $3.39   
2                                               [Apple]     $3.39   
3                                                   NaN     $2.39   
4     [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98   
...                                                 ...        ...  
4617  [Fresh Tomato Salsa, [Rice, Black Beans, Sour ...    $11.75   
4618  [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese...    $11.75   
4619  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...    $11.25   
4620  [Fresh Tomato Salsa, [Fajita Vegetables, Lettu...     $8.75   
4621  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...     $8.75   

[4622 rows x 5 columns]

### Step 4. See the first 10 entries

In [4]:
a.head(10)

order_id  quantity                              item_name  \
0         1         1           Chips and Fresh Tomato Salsa   
1         1         1                                   Izze   
2         1         1                       Nantucket Nectar   
3         1         1  Chips and Tomatillo-Green Chili Salsa   
4         2         2                           Chicken Bowl   
5         3         1                           Chicken Bowl   
6         3         1                          Side of Chips   
7         4         1                          Steak Burrito   
8         4         1                       Steak Soft Tacos   
9         5         1                          Steak Burrito   

                                  choice_description item_price  
0                                                NaN     $2.39   
1                                       [Clementine]     $3.39   
2                                            [Apple]     $3.39   
3                                                NaN     $2.39   
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98   
5  [Fresh Tomato Salsa (Mild), [Rice, Cheese, Sou...    $10.98   
6                                                NaN     $1.69   
7  [Tomatillo Red Chili Salsa, [Fajita Vegetables...    $11.75   
8  [Tomatillo Green Chili Salsa, [Pinto Beans, Ch...     $9.25   
9  [Fresh Tomato Salsa, [Rice, Black Beans, Pinto...     $9.25

### Step 5. What is the number of observations in the dataset?

In [5]:
# a.shape[0] is simply the total number of rows in your DataFrame.
# You are asking for the first element of the tuple that a.shape returns.
a.shape[0]

4622

In [34]:
# a.shape gives you the dimensions of the DataFrame as a tuple (rows, columns).
a.shape

(4622, 5)

In [35]:
# a.dtypes

# a.tail(15)

a.size   #row*columns

23110

#### show you the difference between a.shape[0] vs len(a) (they look similar but can differ in some cases)?


#### Step 6. What is the number of columns in the dataset?

In [36]:
a.shape[1]

5

### Step 7. Print the name of all the columns.

In [37]:
a.columns

Index(['order_id', 'quantity', 'item_name', 'choice_description',
       'item_price'],
      dtype='object')

### Step 8. How is the dataset indexed?

In [38]:
a.index

RangeIndex(start=0, stop=4622, step=1)

### Step 9. Which was the most-ordered item?

In [39]:
# If no quantity column → use a.value_counts().idxmax()
a.value_counts().idxmax() #this is wrong for this dataframe

(np.int64(724), np.int64(1), 'Canned Soft Drink', '[Coke]', '$1.25 ')

In [40]:
# a.groupby("item_name")["quantity"].sum().idxmax()
a.groupby('item_name')['quantity'].sum().idxmax()

'Chicken Bowl'

In [41]:
a.groupby('item_name')["quantity"].sum().sort_values(ascending=False).head(10)

item_name
Chicken Bowl                    761
Chicken Burrito                 591
Chips and Guacamole             506
Steak Burrito                   386
Canned Soft Drink               351
Chips                           230
Steak Bowl                      221
Bottled Water                   211
Chips and Fresh Tomato Salsa    130
Canned Soda                     126
Name: quantity, dtype: int64

* result.max() → 761 (the highest quantity)

* result.idxmax() → "Chicken Bowl" (the item with that highest quantity)

Summary:

* If no quantity column → use .value_counts().idxmax()

* If quantity exists → use .groupby('item')['quantity'].sum().idxmax()

### Would you like me to also show you the same for **top 3 most-ordered items** (instead of only the max one)?

In [42]:
top3= a.groupby("item_name")["quantity"].sum().sort_values(ascending=False).head(3)
print(top3)

item_name
Chicken Bowl           761
Chicken Burrito        591
Chips and Guacamole    506
Name: quantity, dtype: int64


### Step 10. For the most-ordered item, how many items were ordered?

In [43]:
a.groupby('item_name')['quantity'].sum().idxmax()

'Chicken Bowl'

### Step 11. What was the most ordered item in the choice_description column?

In [44]:
a['choice_description'].value_counts().idxmax()

'[Diet Coke]'

### Step 12. How many items were orderd in total?

In [45]:
a['quantity'].sum()

np.int64(4972)

### Step 13. Turn the item price into a float

In [46]:
a.head(5)

order_id  quantity                              item_name  \
0         1         1           Chips and Fresh Tomato Salsa   
1         1         1                                   Izze   
2         1         1                       Nantucket Nectar   
3         1         1  Chips and Tomatillo-Green Chili Salsa   
4         2         2                           Chicken Bowl   

                                  choice_description item_price  
0                                                NaN     $2.39   
1                                       [Clementine]     $3.39   
2                                            [Apple]     $3.39   
3                                                NaN     $2.39   
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98

In [47]:
# a['item_price'] = (
    # a['item_price']
    # .astype(str)                         # ensure it's a string
    # .str.replace('[^0-9.]', '', regex=True)  # remove everything except digits and dot
    # .astype(float)                       # convert to float
# )

In [48]:
a['item_price'] = (
    a['item_price']
    .str.replace('$', '', regex=True)   # remove $
    .str.strip()                        # remove spaces before/after
    .astype(float)                      # convert to float
)

ValueError: could not convert string to float: '$2.39'

In [ ]:
a['item_price'] = a['item_price'].str.replace('$', '', regex=True).str.strip().astype(float)

In [ ]:
a.head(5)

#### Step 13.a. Check the item price type

In [ ]:
a['item_price'].dtypes

#### Step 13.b. Create a lambda function and change the type of item price

In [ ]:
import re
a['item_price']=a['item_price'].apply(lambda x:float(re.sub('[^0-9.]','',x)))

#### re stands for regular expressions — it’s a Python built-in module used for pattern matching and text manipulation (finding, replacing, or extracting parts of strings).

#### Step 13.c. Check the item price type

In [ ]:
a['item_price'].dtypes

### Step 14. How much was the revenue for the period in the dataset?

In [ ]:
a['revenue']=a['quantity']*a['item_price']

In [ ]:
total_revenue1=(a['quantity']*a['item_price']).sum()
total_revenue1

In [ ]:
total_revenue=a['revenue'].sum()
total_revenue

### Step 15. How many orders were made in the period?

In [ ]:
a.columns

In [ ]:
num_order=a['order_id'].nunique()
num_order

### Step 16. What is the average revenue amount per order?

In [ ]:
avg_rev_per_order=total_revenue/num_order

In [ ]:

avg_rev_per_order

### Step 17. How many different items are sold?

In [ ]:
num_diff_items=a['item_name'].nunique()

In [ ]:
num_diff_items